File: exercise/C#/day13/SantaMarket/Model/ChristmasElf.cs
namespace SantaMarket.Model
{
    public class ChristmasElf(ISantamarketCatalog catalog)
    {
        private readonly Dictionary<Product, Offer> _offers = new();

        public void AddSpecialOffer(SpecialOfferType offerType, Product product, double argument)
            => _offers[product] = new Offer(offerType, product, argument);

        public Receipt ChecksOutArticlesFrom(ShoppingSleigh thesleigh)
        {
            var receipt = new Receipt();
            var productQuantities = thesleigh.Items();

            foreach (var pq in productQuantities)
            {
                var p = pq.Product;
                var quantity = pq.Quantity;
                var unitPrice = catalog.GetUnitPrice(p);
                var price = quantity * unitPrice;
                receipt.AddProduct(p, quantity, unitPrice, price);
            }

            thesleigh.HandleOffers(receipt, _offers, catalog);

            return receipt;
        }
    }
}

File: exercise/java/day13/src/main/java/santamarket/model/ChristmasElf.java
package santamarket.model;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class ChristmasElf {
    private final SantamarketCatalog catalog;
    private final Map<Product, Offer> offers = new HashMap<>();

    public ChristmasElf(SantamarketCatalog catalog) {
        this.catalog = catalog;
    }

    public void addSpecialOffer(SpecialOfferType offerType, Product product, double argument) {
        offers.put(product, new Offer(offerType, product, argument));
    }

    public Receipt checksOutArticlesFrom(ShoppingSleigh thesleigh) {
        Receipt receipt = new Receipt();
        List<ProductQuantity> productQuantities = thesleigh.getItems();
        for (ProductQuantity pq : productQuantities) {
            Product p = pq.getProduct();
            double quantity = pq.getQuantity();
            double unitPrice = catalog.getUnitPrice(p);
            double price = quantity * unitPrice;
            receipt.addProduct(p, quantity, unitPrice, price);
        }
        thesleigh.handleOffers(receipt, offers, catalog);

        return receipt;
    }
}
